### PR TITLE
perf: use inheritance to improve our code base

### LINKED ISSUE #40: Use inheritance to improve our code base
As @lqmanh mentioned [here](https://github.com/supabase-community/postgrest-py/pull/30#issuecomment-959742469), inheritance is the way to go but #30 was already too big to include.

Therefore, we should open a PR refactoring the lib using inheritance.

### PR DESCRIPTION
**Pull request by leynier**
Resolve #40 
Resolve https://github.com/supabase-community/supabase-py/issues/105

### CODE REVIEW

**Review by dreinon — CHANGES_REQUESTED** [review]

File: postgrest_py/base_request_builder.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+
+def process_response(
+    session: Union[AsyncClient, SyncClient],
+    r: Response,
+) -> Tuple[Any, Optional[int]]:
+    count = None
+    try:
+        count_header_match = re.search(
+            "count=(exact|planned|estimated)", session.headers["prefer"]
```
[the remark this answers is not in the document]
Comment (leynier) [reply postgrest_py/base_request_builder.py:90]: This header comes from httpx

File: postgrest_py/__init__.py [hunk 2]
```diff
@@ -8,6 +8,6 @@
 from postgrest_py._sync.request_builder import SyncQueryRequestBuilder  # noqa: F401
 from postgrest_py._sync.request_builder import SyncRequestBuilder  # noqa: F401
 from postgrest_py._sync.request_builder import SyncSelectRequestBuilder  # noqa: F401
-from postgrest_py.config import DEFAULT_POSTGREST_CLIENT_HEADERS  # noqa: F401
```
[the remark this answers is not in the document]
Comment (leynier) [reply postgrest_py/__init__.py:11]: Ok 👍🏼

File: postgrest_py/base_request_builder.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+    def match(self, query: Dict[str, Any]):
+        updated_query = None
+        for key in query:
+            value = query.get(key, "")
+            updated_query = self.eq(key, value)
+        return updated_query
+
+
+class BaseSelectRequestBuilder(BaseFilterRequestBuilder):
```
[the remark this answers is not in the document]
Comment (leynier) [reply postgrest_py/base_request_builder.py:193]: This pull request does not change the behavior itself. On the other hand, `postgrest-js` allows calling order, then the eq filter, and then order again. So I propose to resolve that issue in another pull request.

**Review by J0 — APPROVED** [review]

File: postgrest_py/base_request_builder.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dreinon) [reply postgrest_py/base_request_builder.py:210]: Okay, sure! :+1:

**Review by dreinon — APPROVED** [review]

**Review by anand2312 — APPROVED** [review]
Whoops, I'd looked through this earlier, forgot to approve. LGTM 🚀

**Comment by lqmanh** [comment]
LGTM

**Comment by J0** [comment]
Hey @lqmanh,

Sorry to bother you again but we were hoping to publish a new release with these changes when we realized that we don't quite have access to publish to `postgrest-py` https://pypi.org/project/postgrest-py/ on PyPI. Wondering if you could grant us permissions?

My handle is `joellee` and @dreinon 's handle is `dreinon`

**Comment by lqmanh** [comment]
@J0 OK I'll grant you permission

**Pull request merged**